services/route_manager/src/main/java/com/futurewei/alcor/route/controller/NeutronRouterController.java

```diff
@@ -0,0 +1,375 @@
+        String subnetId = resource.getSubnetId();
+
+        RouterInterfaceResponse routerInterfaceResponse = this.neutronRouterService.addAnInterfaceToNeutronRouter(projectid, portId, subnetId, routerid);
+
+        return routerInterfaceResponse;
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/remove_router_interface"})
+    @DurationStatistics
+    public RouterInterfaceResponse removeInterfaceToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RouterInterfaceRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        String portId = resource.getPortId();
+        String subnetId = resource.getSubnetId();
+
+        RouterInterfaceResponse routerInterfaceResponse = this.neutronRouterService.removeAnInterfaceToNeutronRouter(projectid, portId, subnetId, routerid);
+
+        return routerInterfaceResponse;
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/add_extra_routes"})
+    @DurationStatistics
+    public RoutesToNeutronWebResponse addRoutesToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RoutesToNeutronWebRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        RoutesToNeutronRouterRequestObject router = resource.getRouter();
+        if (router == null) {
+            return new RoutesToNeutronWebResponse();
+        }
+
+        RoutesToNeutronWebResponse routesToNeutronWebResponse = this.neutronRouterService.addRoutesToNeutronRouter(routerid, router);
+
+        return routesToNeutronWebResponse;
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/remove_extra_routes"})
+    @DurationStatistics
+    public RoutesToNeutronWebResponse removeRoutesToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RoutesToNeutronWebRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        RoutesToNeutronRouterRequestObject router = resource.getRouter();
+        if (router == null) {
+            return new RoutesToNeutronWebResponse();
+        }
+
+        RoutesToNeutronWebResponse routesToNeutronWebResponse = this.neutronRouterService.removeRoutesToNeutronRouter(routerid, router);
+
```

Comment:
@kevin-zhonghao Reserve the place to call PM for routing rule updating.

Reply:
Sure